src/cockpit/router.py

```diff
@@ -94,6 +95,7 @@ def send_channel_message(self, channel: str, **kwargs: JsonDocument) -> None:
     def send_channel_control(self, channel, command, **kwargs: JsonDocument) -> None:
         self.router.write_control(channel=channel, command=command, **kwargs)
         if command == 'close':
+            self.router.endpoints[self].remove(channel)
```

Comment:
.. and removal even more so. It's very plausible that the Router wants to do some further checks or cleanups on this. This should be a method, not bare "remote" state hacking.

Reply:
see above.  We can discuss this point further.

Reply:
My main concern is that there should be something special happening if `channel` does not exist in the current endpoints map, or this removes the last channel. Plus perhaps some debug logs?

Reply:
We'll throw an exception in that case.

It's actually a good point — this could happen in response to erroneous messages arriving from outside of the router (ie: from a broken peer).  I'll rethink this a bit.